synapse/storage/databases/main/events_worker.py

```diff
@@ -736,35 +736,56 @@ async def get_stripped_room_state_from_event_context(
             for e in state_to_include.values()
         ]
 
-    def _do_fetch(self, conn: Connection) -> None:
+    async def _do_fetch(self) -> None:
+        """Services requests for events from the `_event_fetch_list` queue."""
+        try:
+            await self.db_pool.runWithConnection(self._do_fetch_txn)
+        except BaseException as e:
+            with self._event_fetch_lock:
+                self._event_fetch_ongoing -= 1
+                event_fetch_ongoing_gauge.set(self._event_fetch_ongoing)
+
+                if self._event_fetch_ongoing == 0 and self._event_fetch_list:
+                    # We are the last remaining fetcher and we have just failed.
+                    # Fail any outstanding event fetches, since no one else will process
+                    # them.
+                    failed_event_list = self._event_fetch_list
+                    self._event_fetch_list = []
+                else:
+                    failed_event_list = []
+
+            for _, deferred in failed_event_list:
+                if not deferred.called:
+                    with PreserveLoggingContext():
+                        deferred.errback(e)
```

Comment:
This is still wrong, in a different way. If `runWithConnection` raises *after* `_do_fetch_txn` successfully exits, we will decrement `_event_fetch_ongoing` twice.

On the other hand, if we move the decrement out of `_event_fetch_ongoing`, then there is a race:
  1. All fetchers have decided to terminate
  2. An event fetch is queued up, but doesn't start a new fetcher
  3. All fetchers terminate, and that event fetch fails.